### DISCUSSION TITLE
Use a custom tfhub URL as model

### DISCUSSION THREAD

**Discussion by mautenr** [BenWiseman/sentiment.ai#9] [Q&A]
Hi! 

With sentiment.ai you created a really powerful tool for analyzing human language! Thanks für the great work!

I wanted to test the possibility of using a custom tfhub URL as the embedded language model. So I initialized the model using:

`init_sentiment.ai(model = "https://tfhub.dev/google/wiki40b-lm-de/1")`

It seemed to work, but when I'm calculating the sentiment score, R is giving me an error:

`sentiment_score(strategies$strat)
Model Running ...
  |                                                                                                               |   0%
Fehler in sentiment.ai::sentiment.env$embed(as_py_list(text[this_inds])) : 
  Versuch eine Nicht-Funktion anzuwenden`

The error message can be translated in: 
"Error in sentiment.ai::sentiment.env$embed(as_py_list(text[this_inds])) : 
Try to apply a non-function"

Am I missing something? It would be great to use a custom URL, because I wanted to test the package with some German texts (and compare the tfhub model with the available multi model). 

Thanks in advance!
Miriam

**Comment by swnydick** [comment]
I'm pretty sure that the problem is that "https://tfhub.dev/google/wiki40b-lm-de/1" is a hub module rather than simply text embeddings. I'm sure that that module includes embeddings, but the assumption is that the embeddings can be loaded using tensorflow_hub load and then applied to text directly, e.g.:

```
embed = hub.load(...)
result   = embed(...)
```

Your model requires a more complicated application and is therefore beyond the scope of the current functionality, although this certainly might be added in the future.

**Reply by BenWiseman** [reply]
So, while you can pass in a custom URL, the sentiment scoring is trained on USE embeddings, so we'd also need to train a new model for those embeddings. 

Part of the package design is to allow context specific/alternative models, however that seems like it's require more dev time than I'm likely to have for a while (plus my German is just a few classes in highschool, so I'm not well suited to QA that!). 

If you are able to work out the code to make the module work (looks like a python wrapper function could do the trick) and have a good german language training set, a DE specific model would be a great addition and I'd love to add it. I could, at some point in the future, take a look at making the embedding module work, but, without a DE training set, I wouldn't be able to do anything with the embeddings.

**Comment by mautenr** [comment]
Hey Steven, hey Ben!

Thanks for the explanations! Unfortunately, I'm new to this kind of stuff, so currently it's learning by doing :-) 
But I'll talk to some colleagues, maybe we can set up a good German training set. I think, this would be fantastic addition, because most of the time, sentiment analysis tools can't handle German at all.
